### PR TITLE
Add --reporter parameter

### PR DESCRIPTION
**Pull request by sunshinejr**
Adds optional reporter parameter with (currently) two options: "raw" and "json". Raw is the current one, while "json" returns a json string (useful to libraries using the output of linters, e.g. Danger).

As there is no doc around contributing to this repo: If this is something you don't want to see, just let me know and I can work on my own fork. Cheers!

### CODE REVIEW

**Comment by sunshinejr** [comment]
Hey @ikonst @keith, any answer on this one would be much appreciated.

**Comment by ikonst** [comment]
Hey @sunshinejr, sorry for the delay. I've missed your pull request originally, so thanks for tagging me. We also use Danger for the Lyft iOS apps, so your change is very welcome. By the way, the so-called "raw" output was actually designed for ingest by another one of our tools, https://github.com/lyft/linty_fresh.

Let me review your code and hopefully merge it.

File: xiblint/xibcontext.py [hunk 1]
```diff
@@ -29,11 +30,17 @@ def error(self, element, message, *args):
...
-            message.format(*args),
-            self.rule_name,
-        ))
+            message.format(*args)
+        )
+        new_error_dict = {
+            "file": self.path,
+            "line": element.line,
+            "error": new_error,
+            "rule": self.rule_name
```
Comment (ikonst) [inline xiblint/xibcontext.py]: nit - trailing comma (should add a linter for this :)

File: xiblint/__main__.py [hunk 2]
```diff
@@ -47,25 +52,40 @@ def main():
...
+
+
+def print_errors(errors, reporter):
+    if reporter == "raw":
+        for error_dict in errors:
+            print("{}:{}: error: {} [rule: {}]".format(
+                error_dict["file"],
+                error_dict["line"],
+                error_dict["error"],
+                error_dict["rule"]
```
Comment (ikonst) [inline xiblint/__main__.py]: nit - trailing comma

File: xiblint/__main__.py [hunk 3]
```diff
@@ -47,25 +52,40 @@ def main():
...
     # Process paths
     #
-    success = True
+    errors = []
     for path in config.include_paths:
         for root, _, files in os.walk(path):
             for file_path in [os.path.join(root, file) for file in files]:
                 checkers = config.checkers(file_path)
-                success = process_file(file_path, checkers) and success
+                errors = errors + process_file(file_path, checkers, reporter)
```
Comment (ikonst) [inline xiblint/__main__.py]: nit - `errors = errors +` —> `errors +=`

File: xiblint/__main__.py [hunk 4]
```diff
@@ -31,7 +32,11 @@ def main():
     )
     parser.add_argument("-v", "--version", action="version",
                         version=__version__)
-    parser.parse_args()
+    parser.add_argument("--reporter", choices=("raw", "json"),
+                        default="raw",
+                        help="custom reporter to use (optional)")
```
Comment (ikonst) [inline xiblint/__main__.py]: argparse arguments are always optional, so perhaps we can omit the "(optional)". The default value "raw" would be indicated as part of the help text anyway.

File: xiblint/__main__.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 import argparse
 import os
 import sys
+import json
```
Comment (ikonst) [inline xiblint/__main__.py]: let's move this after `argparse`. I should add `isort` as a linter...

File: xiblint/__main__.py [hunk 6]
```diff
@@ -31,7 +32,11 @@ def main():
     )
     parser.add_argument("-v", "--version", action="version",
                         version=__version__)
-    parser.parse_args()
+    parser.add_argument("--reporter", choices=("raw", "json"),
+                        default="raw",
+                        help="custom reporter to use (optional)")
+    args = parser.parse_args()
+    reporter = args.reporter
```
Comment (ikonst) [inline xiblint/__main__.py]: After you assign this variable, you're never changing it, so perhaps directly using `args.reporter` will make it clearer that it's coming from the args.

File: xiblint/__main__.py [hunk 7]
```diff
@@ -47,25 +52,40 @@ def main():
...
         for root, _, files in os.walk(path):
             for file_path in [os.path.join(root, file) for file in files]:
                 checkers = config.checkers(file_path)
-                success = process_file(file_path, checkers) and success
+                errors = errors + process_file(file_path, checkers, reporter)
 
-    sys.exit(0 if success else 1)
+    print_errors(errors, reporter)
 
+    sys.exit(0 if errors.count == 0 else 1)
```
Comment (ikonst) [inline xiblint/__main__.py]: Did you mean `len(errors) == 0`? ;)

File: xiblint/xibcontext.py [hunk 8]
```diff
@@ -29,11 +30,17 @@ def error(self, element, message, *args):
...
-            self.path,
-            element.line,
+
+        new_error = "{}{}{}".format(
             "{}: ".format(object_id) if object_id else '',
             "'{}': ".format(moniker) if moniker else '',
-            message.format(*args),
-            self.rule_name,
-        ))
+            message.format(*args)
```
Comment (ikonst) [inline xiblint/xibcontext.py]: trailing comma

File: xiblint/__main__.py (hunk 7 above)
Comment (sunshinejr) [reply xiblint/__main__.py]: I think I couldn't get this one right and was hoping that you could help me with it in a PR but forgot to comment it 🤦‍♂️ thanks 😄

File: xiblint/__main__.py (hunk 3 above)
Comment (sunshinejr) [reply xiblint/__main__.py]: ye, habbit from Swift 😅

**Comment by sunshinejr** [comment]
@ikonst thanks for the review! I've added a commit with suggested fixes and tested it locally and it still works 😄

File: xiblint/__main__.py [hunk 9]
```diff
@@ -47,25 +51,40 @@ def main():
...
         for root, _, files in os.walk(path):
             for file_path in [os.path.join(root, file) for file in files]:
                 checkers = config.checkers(file_path)
-                success = process_file(file_path, checkers) and success
+                errors += process_file(file_path, checkers, args.reporter)
 
-    sys.exit(0 if success else 1)
+    print_errors(errors, args.reporter)
 
+    sys.exit(len(errors) == 0)
```
Comment (ikonst) [inline xiblint/__main__.py]: That looks wrong. You want to exit with zero for success and non-zero for failure. You can also simply say:
```
sys.exit(1 if errors else 0)
```

Comment (sunshinejr) [reply xiblint/__main__.py]: Oh, my mistake. Was so ashamed of this line that I quickly copied/pasted this one and wanted to forget about it 😓 Now tested the statement using the python interpreter and it works for an empty array as well 🤔 Didn't know that it worked that way as well ;-)

**Comment by ikonst** [comment]
Thanks again @sunshinejr. Excited to see this is useful outside of our company!

**Pull request merged**